### ISSUE TITLE
feat: LLM inference integration — inter-token GPU dispatch

### ISSUE THREAD

**Issue by kavanaghpatrick** [kavanaghpatrick/emojiasm#23]
# LLM Inference Integration — PRD

## Problem

EmojiASM programs should execute on the same GPU that runs LLM inference, with results feeding back into the next prompt without leaving unified memory. This is the "GPU-maxxed" endgame.

### Dispatch Architecture (from inference integration agent)

- **Architecture A: Inter-token dispatch (RECOMMENDED)** — kernel encoded into same `MTL::CommandBuffer` as inference, ~1.5us dispatch overhead (measured M4 Pro)
- **Architecture B: Sidecar stream** — Apple GPU hardware partitions ~50% of cores per active queue. Steals half GPU from inference. Not recommended.
- **Architecture C: Separate process** — Same 50/50 split problem plus process coordination. Only for long-running EmojiASM.
- **No compute QoS** on Apple GPU — no way to deprioritize sidecar kernel

### Dispatch Latency (measured)

| Method | Latency |
|--------|---------|
| Intra-command-buffer | ~1.5us |
| CPU-initiated dispatch | ~120us |
| MTLSharedEvent sync | ~21us |
| First dispatch (GPU cold) | ~2.5ms |

### MLX Integration Points (from MLX kernel agent)

- **Lazy evaluation** — `mx.eval()` triggers DFS walk, builds topological tape, dispatches all kernels
- **Auto-fence sync** — Metal fences between dependent operations, no manual barriers
- **MLX-LM `logits_processors`** — callable receiving `(tokens, logits)`, can wrap EmojiASM kernel
- **Read-only inputs** — custom kernel inputs are `const device` (MLX issue #2547)

### Zero-Copy Pipeline (from unified memory agent)

- **KB #140** — `storageModeShared` = truly zero-copy, hardware-managed cache coherence via SLC
- **SLC coherency latency** — ~234ns at SLC hit
- **No manual flush** needed — coherency guaranteed at command buffer boundaries

## Architecture

```
┌──────────────────────────────────────────────────────┐
│                    Apple Silicon GPU                  │
│                                                      │
│  ┌──────────────┐  same cmd buf  ┌────────────────┐  │
│  │ MLX Inference │──────────────▶│ EmojiASM Kernel │  │
│  │ (forward pass)│  ~1.5us fence │ (N VM threads)  │  │
│  │ logits tensor │               │ results tensor  │  │
│  └──────┬───────┘               └────────┬─────────┘ │
│         │      unified memory (zero-copy) │          │
│         └────────────────────────────────┘           │
└──────────────────────────────────────────────────────┘
                      │
          mx.eval() triggers both
                      │
              ┌───────▼────────┐
              │  Agent Loop    │
              │  (Python)      │
              └────────────────┘
```

## API

```python
# emojiasm/inference.py

from .gpu import gpu_run, gpu_available
from .parser import parse
from .bytecode import compile_to_bytecode, gpu_tier

class EmojiASMTool:
    """LLM tool that executes EmojiASM programs on GPU.

    Designed to be called between token generation steps.
    Results stay in GPU memory as mx.array for zero-copy
    feedback to the next prompt.
    """

    def __init__(self, max_instances: int = 10_000):
        self.max_instances = max_instances
        self._kernel = None  # lazy init

    def execute(self, source: str, n: int = 1) -> dict:
        """Parse and execute an EmojiASM program.

        Args:
            source: EmojiASM source code (from LLM output)
            n: Number of parallel instances

        Returns:
            dict with results, stats, mode (gpu/cpu), timing
        """
        program = parse(source)
        tier = gpu_tier(program)

        if tier <= 2 and n >= 256 and gpu_available():
            return gpu_run(program, n=n)
        else:
            from .agent import run_agent_mode
            return run_agent_mode(program, runs=n)

    def as_logits_processor(self):
        """Wrap as MLX-LM logits processor for inline GPU dispatch."""
        # Returns callable(tokens, logits) -> logits
        # Detects EmojiASM tool call tokens in output
        # Dispatches kernel, injects results
        ...
```

## MLX-LM Integration Example

```python
import mlx_lm
from emojiasm.inference import EmojiASMTool

tool = EmojiASMTool(max_instances=10_000)

# Generate with inline EmojiASM execution
response = mlx_lm.generate(
    model, tokenizer, prompt,
    logits_processors=[tool.as_logits_processor()],
)

# Or manual tool-call loop:
while True:
    output = mlx_lm.generate(model, tokenizer, prompt)
    if "<emojiasm>" in output:
        source = extract_emojiasm(output)
        result = tool.execute(source, n=10_000)
        prompt += f"\nResult: {result['stats']}"
    else:
        break
```

## Tests Required

- `EmojiASMTool.execute()` routes to GPU for n>=256 numeric programs
- `EmojiASMTool.execute()` routes to CPU for n<256 or tier-3 programs
- GPU results match CPU results for same program
- End-to-end: parse → bytecode → GPU → results in one call
- Graceful fallback when MLX not available
- JSON output compatible with agent-mode schema
- `as_logits_processor()` returns callable with correct signature

## Dependencies

**Depends on:** #19 (Bytecode), #20 (MSL Kernel), #21 (MLX Backend)
**Blocks:** Nothing — this is the capstone
